lib/upipe-modules/upipe_audio_merge.c

```diff
@@ -0,0 +1,638 @@
+        upipe_warn(upipe, "Got new uref before last one was output!");
+    }
+    upipe_audio_merge_sub->uref = uref;
+
+    // produce output, if we have all urefs
+    upipe_audio_merge_produce_output(upipe_audio_merge_to_upipe(upipe_audio_merge), upump_p);
+}
+
+/** @This frees a upipe.
+ *
+ * @param upipe description structure of the pipe
+ */
+static void upipe_audio_merge_sub_free(struct upipe *upipe)
+{
+    struct upipe_audio_merge_sub *upipe_audio_merge_sub =
+                              upipe_audio_merge_sub_from_upipe(upipe);
+
+    uref_free(upipe_audio_merge_sub->uref);
+    uref_free(upipe_audio_merge_sub->flow_def);
+
+    upipe_throw_dead(upipe);
+
+    upipe_audio_merge_sub_clean_sub(upipe);
+    upipe_audio_merge_sub_clean_urefcount(upipe);
+    upipe_audio_merge_sub_free_void(upipe);
+}
+
+/** @internal @This initializes the output manager for an audio_merge sub pipe.
+ *
+ * @param upipe description structure of the pipe
+ */
+static void upipe_audio_merge_init_sub_mgr(struct upipe *upipe)
+{
+    struct upipe_audio_merge *upipe_audio_merge =
+                              upipe_audio_merge_from_upipe(upipe);
+    struct upipe_mgr *sub_mgr = &upipe_audio_merge->sub_mgr;
+    sub_mgr->refcount = upipe_audio_merge_to_urefcount_real(upipe_audio_merge);
+    sub_mgr->signature = UPIPE_AUDIO_MERGE_INPUT_SIGNATURE;
+    sub_mgr->upipe_alloc = upipe_audio_merge_sub_alloc;
+    sub_mgr->upipe_input = upipe_audio_merge_sub_input;
+    sub_mgr->upipe_control = upipe_audio_merge_sub_control;
+    sub_mgr->upipe_mgr_control = NULL;
+}
+
+/** @internal @This allocates an audio_merge pipe.
+ *
+ * @param mgr common management structure
+ * @param uprobe structure used to raise events
+ * @param signature signature of the pipe allocator
+ * @param args optional arguments
+ * @return pointer to upipe or NULL in case of allocation error
+ */
+static struct upipe *upipe_audio_merge_alloc(struct upipe_mgr *mgr,
+                                             struct uprobe *uprobe,
+                                             uint32_t signature, va_list args)
+{
+    struct uref *flow_def;
+    struct upipe *upipe = upipe_audio_merge_alloc_flow(mgr,
+                            uprobe, signature, args, &flow_def);
+    if (unlikely(upipe == NULL))
+        return NULL;
+
+    struct upipe_audio_merge *upipe_audio_merge =
+                              upipe_audio_merge_from_upipe(upipe);
+    upipe_audio_merge_init_urefcount(upipe);
+    urefcount_init(upipe_audio_merge_to_urefcount_real(upipe_audio_merge),
+                   upipe_audio_merge_free);
+
+    upipe_audio_merge_init_upump_mgr(upipe);
+    upipe_audio_merge_init_upump(upipe);
+    upipe_audio_merge_init_output(upipe);
+    upipe_audio_merge_init_sub_mgr(upipe);
+    upipe_audio_merge_init_sub_inputs(upipe);
+    upipe_audio_merge_init_ubuf_mgr(upipe);
+
+    upipe_audio_merge->sub_flow_def = NULL;
+
+    upipe_audio_merge_store_flow_def(upipe, flow_def);
```

Comment:
This helper will throw an event, so you'll have to do that after the `upipe_throw_ready(upipe);`